core/src/sumeragi/fault.rs

```diff
@@ -1305,500 +872,511 @@ pub fn run_sumeragi_main_loop<F>(
-                        }
-                    }
-                }
+    iroha_logger::info!("Initializing iroha using the genesis block.");
 
-                continue;
+    state_machine_guard.latest_block_height = 0;
+    state_machine_guard.latest_block_hash = Hash::zeroed().typed();
+
+    let transactions = genesis_network.transactions.clone();
+    // Don't start genesis round. Instead just commit the genesis block.
+    if transactions.is_empty() {
+        panic!("Genesis transaction set contains no valid transactions");
+    } else {
+        let block = PendingBlock::new(transactions, Vec::new())
+            .chain_first_with_genesis_topology(state_machine_guard.current_topology.clone());
+
+        {
+            info!(block_hash = %block.hash(), "Publishing genesis block.");
+
+            let block = block.validate(&sumeragi.transaction_validator, &state_machine_guard.wsv);
+
+            info!(
+                peer_role = ?state_machine_guard.current_topology.role(&sumeragi.peer_id),
+                block_hash = %block.hash(),
+                "Created a block to commit.",
+            );
+            for event in Vec::<Event>::from(&block) {
+                trace!(?event);
+                drop(sumeragi.events_sender.send(event));
+            }
+            let signed_block = block
+                .sign(sumeragi.key_pair.clone())
+                .expect("Sign genesis block.");
+            {
+                sumeragi.broadcast_msg(
+                    BlockCommitted::from(signed_block.clone()),
+                    &state_machine_guard.current_topology,
+                );
+                block_commit(sumeragi, signed_block, &mut state_machine_guard);
             }
         }
     }
 }
 
-impl<F: FaultInjection> Debug for SumeragiWithFault<F> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Sumeragi")
-            .field("public_key", &self.key_pair.public_key())
-            // .field("network_topology", &self.topology) TODO FIX
-            .field("peer_id", &self.peer_id)
-            //.field("voting_block", &self.voting_block)
-            .finish()
+fn sumeragi_init_listen_for_genesis<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    incoming_message_receiver: &mut mpsc::Receiver<Message>,
+) where
+    F: FaultInjection,
+{
+    use std::sync::mpsc::TryRecvError;
+
+    trace!("Start listen for genesis.");
+    let mut state_machine_guard = sumeragi
+        .sumeragi_state_machine_data
+        .lock()
+        .expect("take lock");
+
+    if !state_machine_guard.current_topology.is_consensus_required() {
+        panic!(
+            "How am I supposed to receive a genesis block if I am the only peer in the network?"
+        );
+    }
+    {
+        *sumeragi
+            .latest_block_hash_for_use_by_block_sync
+            .lock()
+            .expect("push hash to block sync for genesis") = Hash::zeroed().typed();
+    }
+
+    loop {
```

Comment:
Oh, i was thinking about simmilar approah: instead of just `finite state machine` introduce [`hierarchical finite state machine`](https://web.stanford.edu/class/cs123/lectures/CS123_lec08_HFSM_BT.pdf).